torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+
+    def _get_gradient_predivide_factor(self, world_size: int) -> float:
+        factor: int = 1
+        while world_size % factor == 0 and world_size / factor > factor:
+            factor *= 2
+        return float(factor)
+
+    def _cast_buffers(
+        self, device: Optional[torch.device] = None, memo: Optional[Set] = None
+    ) -> None:
+        """Move all buffers to the given *device*.
+        If *device* are not given, then they will default to
+        ``self.compute_device``. In the
+        case of nested FSDP instances, we will respect the child instance's
+        ``compute_device`` configuration.
+        Args:
+            device (torch.device, Optional):
+                device to cast buffers to (defaults to compute_device)
+            memo (Set, Optional):
+                set of modules that have already been processed
+        """
+        if memo is None:
+            memo = set()
+        for module in self.modules():
+            if module is not self and isinstance(module, FullyShardedDataParallel):
+                # Allow any child FSDP instances to handle their own buffers.
+                module._cast_buffers(device=device, memo=memo)
+            elif module not in memo:
+                memo.add(module)
+                for name, buf in module.named_buffers(recurse=False):
+                    if buf is None:
+                        continue
+                    buf = buf.to(device=device or self.compute_device)
+                    setattr(module, name, buf)
+
+    @torch.no_grad()
+    def _shard_parameters_(self) -> None:
+        """
+        At initialization we wrap a module with full parameters and shard the
+        parameters in-place. Sharding is implemented by viewing each parameter
+        as a 1D Tensor and retaining only a single slice, where the slice size
+        is determined by the number of data parallel workers.
+        After this initial sharding is complete, the user can initialize a
+        ``torch.optim.Optimizer`` in the usual way, i.e.::
+        .. code-block:: python
+            optim = torch.optim.Adam(sharded_module.parameters(), lr=0.0001)
+        The optimizer will see only a single slice of parameters and will thus
+        allocate less memory for optimizer state, avoiding redundancy across
+        data parallel workers.
+        """
+        self.numel_padded_per_param = []
+        for p in self.params:
+            assert not hasattr(p, "_is_sharded")
+            assert p.is_floating_point()
+
+            # Sharding is done only when world_size is larger than 1.
+            p._is_sharded = self.world_size > 1  # type: ignore[attr-defined]
+            p._orig_size = p.size()  # type: ignore[attr-defined]
+
+            if not p._is_sharded:  # type: ignore[attr-defined]
+                self.numel_padded_per_param.append(0)
+                continue
+
+            # Replace p.data with the relevant shard.
+            orig_data = p.data
```

Comment:
You can simply get p's storage here so that you don't need to use `.data`: `orig_storage = p.storage()` and the free_storage function below can take a storage as input.